src/syntaxtree.c

```diff
@@ -158,6 +185,74 @@ static Obj SyntaxTreeDefaultCompiler(Obj result, Expr expr)
     return result;
 }
 
+// TODO: use the following as start of a "reverse compiler"
+static Expr SyntaxTreeDefaultCoder(Obj node)
+{
+
+    Obj  typeString = ElmPRec(node, RNamName("type"));
+    UInt tnum = FIND_TNUM_FROM_STRING(CSTR_STRING(typeString));
+
+    if (tnum == LAST_EXPR_TNUM + 2) {
+        ErrorQuit("Unrecognized syntax tree node type", 0, 0);
+    }
+
+    CompilerT comp = Compilers[tnum];
+
+    if (comp.code != SyntaxTreeDefaultCoder) {
+        return comp.code(node);
+    }
+
+    // compute how many slots we need
+    UInt slots = comp.arity;
+    UInt arity = comp.arity;
+
+    UInt isvararg = comp.arity > 0 && comp.args[comp.arity - 1].argcomp == 0;
+    Obj  vararglist;
+
+    if (isvararg) {
+        arity--;
+        vararglist = ElmPRec(node, RNamName(comp.args[arity].argname));
+        slots = arity + LEN_LIST(vararglist);
+    }
+
+    // reserve space for the statement or expressions
+    Expr expr = NewStatWithProf(tnum, slots * sizeof(Expr), 0);
+
+    UInt i;
+
+    for (i = 0; i < arity; i++) {
+        UInt rnam = RNamName(comp.args[i].argname);
+        Obj  subast = ElmPRec(node, rnam);
+        WRITE_EXPR(expr, i, comp.args[i].argcode(subast));
+    }
+
+    if (isvararg) {
```

Comment:
You know, you don't even need that `isvararg` check here: if it is 0, then arity == slots, and the following loop does not run anyway. Though I guess this might trigger a compiler warning "`vararglist` may be use uninitialized?

Reply:
Indeed. Also the check is cheap, and because we need the `isvararg` anyway, it makes the code easier to read.